Add option for ignoring auto DNS and routes

cathay4t opened the pull request:
Added these options to ipv4 and ipv6 section:

 * ignore-auto-routes:  Ignore all DHCP/autoconf routes(including gateway).
 * ignore-auto-gateway: Ignore the DHCP/autoconf gateway.
 * ignore-auto-dns:     Ignore the DNS provided by DHCP.

EdDev commented on libnmstate/nm/ipv4.py:
```diff
@@ -67,6 +76,9 @@ def get_info(active_connection):
     if ip_profile:
         info['dhcp'] = ip_profile.get_method() == (
             nmclient.NM.SETTING_IP4_CONFIG_METHOD_AUTO)
+        info['ignore-auto-routes'] = ip_profile.props.ignore_auto_routes
```
If it was never set, won't it return `None`?

Should we return these if `dhcp` is disabled?

cathay4t replied:
> If it was never set, won't it return `None`?
The [libnm GI document][1] said it's a bool:

[1]: https://lazka.github.io/pgi-docs/#NM-1.0/classes/SettingIPConfig.html#NM.SettingIPConfig.get_ignore_auto_routes
> 
> Should we return these if `dhcp` is disabled?

That goes to be a debate. Including DHCP/Autoconf options when DHCP/Autoconf is disabled requires extra user changes in yaml/json config when user simply want to enable DHCP(which happens to be most common use case).

EdDev commented on libnmstate/validator.py:
```diff
@@ -92,3 +96,12 @@ def verify_dhcp(state):
                 logging.warning('%s addresses are ignored when '
                                 'dynamic IP is enabled', family)
+
+            if not ip.get('dhcp') and not ip.get('autoconf'):
```
This check is performed before merging with the current state (for an existing iface), therefore, the `dhcp` and `autoconf` may not even be present in the desired state, but existing in the current state.

I am not sure about the implication of moving this verification at a later stage (after the merge).

cathay4t replied:
I think `DesiredStateIsNotCurrentError` holds too much errors, providing different exception
might help user to identify whether this failure is caused by human-error(invalid configure settins) or issue of libnm/nmstate.

If we know something will fail for sure due to invalid config, there is no need to waste on apply-and-rollback.

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -192,6 +193,126 @@ def test_dhcp_for_bond_with_ip_address_and_slave(dhcp_env,
+
+    netapplier.apply(desired_state)
+    assertlib.assert_state(desired_state)
+    # TODO(Gris Ge): Once libnmstate support route query, need to make
```
Assuming the dhcp server does provide a gateway, we can check it with a temporary helper (`ip route` output).

tyll replied:
This seems to be done

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -192,6 +193,126 @@ def test_dhcp_for_bond_with_ip_address_and_slave(dhcp_env,
+    dhcp_cli_desired_state['ipv4']['enabled'] = True
+    dhcp_cli_desired_state['ipv4']['dhcp'] = True
+    dhcp_cli_desired_state['ipv4']['ignore-auto-gateway'] = True
+    dhcp_cli_desired_state['ipv6']['enabled'] = True
```
Why is this enabled if the test is about ipv4?

cathay4t replied:
Once interface up, the IPv6 link-local address will appear, hence we should set ipv6 as enabled also in order to pass the verification stage.

EdDev replied:
That is a bug, and it is not related to this PR changes. If you set it to `false` as a temporary fix, I would understand.

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -192,6 +193,126 @@ def test_dhcp_for_bond_with_ip_address_and_slave(dhcp_env,
+
+    netapplier.apply(desired_state)
+    assertlib.assert_state(desired_state)
+    # TODO(Gris Ge): Once libnmstate support DNS query, need to make
```
Same as before, we could verify this by checking the resolv.conf content using a temporary helper.

tyll replied:
This seems to be done

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -192,6 +193,126 @@ def test_dhcp_for_bond_with_ip_address_and_slave(dhcp_env,
+    dhcp_cli_desired_state['ipv4']['enabled'] = True
+    dhcp_cli_desired_state['ipv4']['dhcp'] = True
+    dhcp_cli_desired_state['ipv4']['ignore-auto-dns'] = True
+    dhcp_cli_desired_state['ipv6']['enabled'] = True
```
Why ipv6?

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -192,6 +193,126 @@ def test_dhcp_for_bond_with_ip_address_and_slave(dhcp_env,
+    dhcp_cli_desired_state['ipv4']['enabled'] = True
+    dhcp_cli_desired_state['ipv4']['dhcp'] = True
+    dhcp_cli_desired_state['ipv4']['ignore-auto-routes'] = True
+    dhcp_cli_desired_state['ipv6']['enabled'] = True
```
Why ipv6?

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -192,6 +193,126 @@ def test_dhcp_for_bond_with_ip_address_and_slave(dhcp_env,
+
+    netapplier.apply(desired_state)
+    assertlib.assert_state(desired_state)
+    # TODO(Gris Ge): Once libnmstate support route query, need to make
```
Is there a flag on the `dhclient` command-line that will should this has been added? Or is this a config file generated by NM? Can we check it as a temporary assertion?

BTW: I would prefer not to sign on the TODO with names, as someone else may take it in the future. The one who wrote the TODO is already saved by `git`.

cathay4t replied:
The nm(version 1.14.5) I am using does not invoke `dhclient` command or other commands.
And I failed to find a simple to verify this option.

OK. I will remove the TODO signing.

EdDev replied:
This is very strange, on CentOS 7.5 `ps -ef | grep dhclient` shows it is running.
What changed?

cathay4t replied:
In manpage of `networkmanager.conf`:

```
dhcp
   This key sets up what DHCP client NetworkManager will use. Allowed values
   are dhclient, dhcpcd, and internal. The dhclient and dhcpcd options require
   the indicated clients to be installed. The internal option uses a built-in
   DHCP client which is not currently as featureful as the external clients.

   If this key is missing, it defaults to internal. It the chosen plugin is not
   available, clients are looked for in this order: dhclient, dhcpcd, internal.
```

I guess since certain version, NetworkManager default internal dhcp client.

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -192,6 +193,126 @@ def test_dhcp_for_bond_with_ip_address_and_slave(dhcp_env,
+    netapplier.apply(desired_state)
+
+
+@pytest.mark.xfail(raises=validator.InvalidDynamicIpOption)
```
`xfail` means it should pass but currently fails and you know about it.
In this case, you expect it to fail, so you should use `with pytest.raises`.

cathay4t replied:
Thanks for pointing that out. Will fix it soon.

tyll commented:
> Added these options to ipv4 and ipv6 section:
> 
> * ignore-auto-routes:  Ignore all DHCP/autoconf routes(including gateway).
> * ignore-auto-gateway: Ignore the DHCP/autoconf gateway.
> * ignore-auto-dns:     Ignore the DNS provided by DHCP.

How about removing the double negation from these settings and just use something like:

`auto-routes`, which defaults to `True` but can be set to `False` to ignore them. IMHO it would fit better in the general nmstate schema.

cathay4t commented:
V2 patch uploaded:

Changes since V1:
 * Renamed the option from `ignore-auto-xxx` to `auto-xxx` as @tyll suggested.
 * Removed the pre-merge check and just remove auto ip option when auto ip disabled.

tyll commented on libnmstate/nm/ipv4.py:
```diff
@@ -27,6 +27,9 @@ def create_setting(config, base_con_profile):
         if setting_ipv4:
             setting_ipv4 = setting_ipv4.duplicate()
             setting_ipv4.clear_addresses()
+            setting_ipv4.props.ignore_auto_routes = False
```
@cathay4t did you see Edy's question?

cathay4t replied:
This is more like the issue #190 
Let's say:

User old config is `dhcp=true` and `ignore-auto-routes`, but new desired config is `dhcp=false`.

Current code does not reset `ignore_auto_routes` property which cause libnm fail as `dhcp=false` but `ignore_auto_routes=True`.

tyll commented on tests/integration/dhcp_test.py:
```diff
@@ -192,6 +192,126 @@ def test_dhcp_for_bond_with_ip_address_and_slave(dhcp_env,
+    assertlib.assert_state(desired_state)
+    # TODO(Gris Ge): Once libnmstate support route query, need to make
+    # sure dhcp route got ignored. And this also require change DHCP server
+    # from dnsmasq to dhcpcd to pushing extra routes.
```
You can push extra routes with ` --dhcp-option=option:classless-static-route,198.51.100.72/30,192.0.2.23` in dnsmasq.

tyll replied:
This seems to be addressed, so the TODO can be removed, can it not?

tyll commented on tests/integration/dhcp_test.py:
```diff
@@ -192,6 +192,126 @@ def test_dhcp_for_bond_with_ip_address_and_slave(dhcp_env,
+    # sure dhcp DNS entries got ignored.
+
+
+def test_ipv4_dhcp_ignore_routes(dhcp_env):
```
It seems that tests that the auto-options actually work are missing.

cathay4t replied:
Could you elaborate on this?

tyll replied:
I mean tests to check if `auto-routes is True` that the routes are actually present in `ip route` output. And the same for the DNS and gateway tests.

tyll commented:
The code itself seems to be working, great! Tested it with dnsmasq and IPv4. It would be great to also test this in the integration tests and to check `/etc/resolv.conf` and `ip route` output do verify that the options were set or ignored.

Another issue, the schema does not match the other option, since we have three options with the same prefix and usually we created a new subtree when we group options. This would mean to do something like:

```
dhcp:
  ip: true
  dns: true
  routes: true
  gateway: true
```

EdDev commented:
> ```
> dhcp:
>   ip: true
> ```

Nice idea.
I would recommend using `enabled` instead of `ip`. We will have to first have a commit that changes current format and then add the others.

cathay4t commented:
V3 patch uploaded.

Changes since V2:
 * Test case expanded to check dns and routes.

tyll commented on tests/integration/dhcp_test.py:
```diff
@@ -31,18 +32,25 @@
 enable-ra
 dhcp-range=2001:db8:1::100,2001:db8:1::fff,64,480h
-""".format(DHCP_SRV_NIC)
+dhcp-option=option:classless-static-route,198.51.100.0/24,192.0.2.1
```
It seems linke you wanted to use the `IPV4_CLASSLESS` constants here.

cathay4t replied:
Done.

cathay4t commented:
V4 patch uploaded.

Changes since V3:
 * Expand the `test_ipv4_dhcp`, `test_ipv6_dhcp_only`, `test_ipv6_dhcp_and_autoconf` test case to check DNS and gateway retrieved from dhcp server.
 * Removed the left-over TODO.

cathay4t commented:
ci test please.

cathay4t commented:
V5 patch uploaded.

Changes since V4:
 * Added new test for switching dhcp from on to off.
 * Use `radvd` to push extra routes via ipv6-ra, and expanded exsiting tests to make sure extra route get set or ignored correctly.

cathay4t commented:
I am uploading new docker image to include radvd. The `docker build` seems takes forever.

cathay4t commented:
The newly build Centos image will fail:

```
Traceback (most recent call last):
  File "/usr/bin/pytest", line 7, in <module>
    from pytest import main
  File "/usr/lib/python2.7/site-packages/pytest.py", line 14, in <module>
    from _pytest.fixtures import fillfixtures as _fillfuncargs
  File "/usr/lib/python2.7/site-packages/_pytest/fixtures.py", line 16, in <module>
    from more_itertools import flatten
  File "/usr/lib/python2.7/site-packages/more_itertools/__init__.py", line 1, in <module>
    from more_itertools.more import *  # noqa
  File "/usr/lib/python2.7/site-packages/more_itertools/more.py", line 329
    def _collate(*iterables, key=lambda a: a, reverse=False):
                               ^
SyntaxError: invalid syntax
```

It's caused by https://github.com/pytest-dev/pytest/issues/4770.

cathay4t commented:
V6 patch uploaded.

Changes since V5:
 * New patch to fix the CentOS CI image on python2 pytest.
 * Massive fixes on dhcp integration test. It passes on my laptop now.

cathay4t commented:
@tyll Ready for review.

tyll commented on tests/integration/dhcp_test.py:
```diff
@@ -102,22 +155,47 @@ def test_ipv6_dhcp_only(dhcp_env):
+    routes = _get_ipv6_main_routes()
+    flag_has_gateway = False
+    flag_has_extra_route = False
+    for route in routes:
```
Since this is used multiple times, it would be good to move this in a method, return `has_gateway` and `has_extra_route`, assert it to True here but also assert it to False when it should not be present. This makes it easier to understand what is happening.

cathay4t replied:
Moved to:

```python
    assert _has_ipv4_dhcp_gateway()
    assert _has_ipv4_classless_route()
```

tyll commented on tests/integration/dhcp_test.py:
```diff
@@ -204,18 +560,74 @@ def _remove_veth_pair():
+        pass
+
+
+def _get_dns_conf():
```
This should be named _get_nameservers() or similar, since it is not the full DNS config.

cathay4t replied:
Done.

tyll commented on tests/integration/dhcp_test.py:
```diff
@@ -204,18 +560,74 @@ def _remove_veth_pair():
+
+def _get_dns_conf():
+    """
+    Return a list of name server string configured in /etc/resolv.conf
```
It is not using /etc/resolv.conf anymore but the constant. How about adding a keyword argument `resolv_conf=RESOLV_CONF_PATH` to document which resolv.conf it is using?

cathay4t replied:
Changed the comment to:

`Return a list of name server string configured in RESOLV_CONF_PATH.`

tyll commented on tests/integration/dhcp_test.py:
```diff
@@ -204,18 +560,74 @@ def _remove_veth_pair():
+        return []
+    return [line for line in output.split('\n') if line]
+
+def _has_ipv6_dns():
```
Maybe rename to `_has_ipv6_nameserver()`.

cathay4t replied:
Done. Renamed to `_has_ipv6_auto_nameserver`.

tyll commented on tests/lib/nm/ipv4_test.py:
```diff
@@ -109,11 +109,14 @@ def test_get_info_with_ipv4_config(NM_mock):
     config_mock.get_addresses.return_value = [address_mock]
     remote_conn_mock = mock.MagicMock()
-    act_con_mock.get_connection().return_value = remote_conn_mock
+    act_con_mock.get_connection.return_value = remote_conn_mock
```
This seems to be an unrelated bugfix. If it is not too much trouble, move it to a separate commit. Also the one in ipv6_test.

tyll commented:
@cathay4t please note that there is a simple flake8 error that makes travis fail.

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -31,21 +33,53 @@
 IPV6_ADDRESS1 = '2001:db8:1::1'
 IPV6_ADDRESS2 = '2001:db8:2::1'
+IPV4_CLASSLESS_ROUTE1 = '198.51.100.0/24'
+IPV4_CLASSLESS_ROUTE_DST1 = '192.0.2.1'
```
It is not clear what this is exactly. Is this the gateway?

cathay4t replied:
No. It's classless route. 

Gateway is `0.0.0.0/2 via <gateway_ip>` when DHCP only provides gateway IP.
Classless route is `198.51.100.0/24 via 192.0.2.1` which require DHCP provides route and next-hop.

EdDev replied:
Usually the format of any route entry is something like: <destination network> via <next hop>.

So perhaps use IPV4_DEST_NET and IPV4_NEXT_HOP or something similar.

tyll replied:
@EdDev's constant's names describe their purpose better, would be nice to change it.

cathay4t replied:
Changed to:
```
IPV4_CLASSLESS_ROUTE_DST_NET1 = '198.51.100.0/24'
IPV4_CLASSLESS_ROUTE_NEXT_HOP1 = '192.0.2.1'
IPV6_CLASSLESS_ROUTE_DST_NET1 = '2001:db8:f::/64'
```

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -89,10 +127,25 @@ def test_ipv4_dhcp(dhcp_env):
 
     netapplier.apply(desired_state)
     assertlib.assert_state(desired_state)
+    time.sleep(5)  # wait dhcp
```
I do not understand this one. Isn't nmstate blocking until the profile is activated? Isn't activation also imply a response from the server with an IP assigned?

cathay4t replied:
The libnm return before IPv4 gateway was applied. No idea the root cause yet.

tyll replied:
Could you please update the comment to say something like `# wait to get resolv.conf updated` or similar.

cathay4t replied:
Done.

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -89,10 +127,25 @@ def test_ipv4_dhcp(dhcp_env):
     assertlib.assert_state(desired_state)
+    time.sleep(5)  # wait dhcp
+    routes = _get_ipv4_main_routes()
+    flag_has_gateway = False
```
Sorry, but this is too 'c' like for my taste.
Having a helper that returns the gateways and another that returns extra routes would have worked out much nicer:
`assert _get_gateways(routes)`
`assert _filter_routes(routes, dst_network, device)`

cathay4t replied:
Thanks. Changed to below:

```python
    assert _has_ipv4_dhcp_dns()
    assert _has_ipv4_dhcp_gateway()
    assert _has_ipv4_classless_route()

```

cathay4t replied:
Thanks. Changed to below:

```python
    assert _has_ipv4_dhcp_nameserver()
    assert _has_ipv4_dhcp_gateway()
    assert _has_ipv4_classless_route()

```

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -204,18 +560,74 @@ def _remove_veth_pair():
         os.unlink(DNSMASQ_CONF_PATH)
     except FileNotFoundError:
         pass
+    except OSError:
```
Cleaner to just do: `except FileNotFoundError, OSError:`

cathay4t replied:
Done.

cathay4t commented on tests/integration/dhcp_test.py:
```diff
@@ -31,21 +33,53 @@
-dhcp-range=2001:db8:1::100,2001:db8:1::fff,64,480h
-""".format(DHCP_SRV_NIC)
-
+dhcp-range=2001:db8:1::100,2001:db8:1::fff,ra-names,slaac,64,480h
```
Done.

cathay4t commented on tests/integration/dhcp_test.py:
```diff
@@ -31,21 +33,53 @@
 DHCP_CLI_NIC = 'dhcpcli'
+DHCP_SRV_IP4 = IPV4_ADDRESS1
+DHCP_SRV_IP6 = IPV6_ADDRESS1
+DHCP_SRV_IP6_NETWORK = '2001:db8:1::/64'
```
Done.

cathay4t commented:
@EdDev 
> Couldn't we solved this one by enforcing the defaults in `sanitize_dhcp_state`?
It also relates to line 43.

If we enforce `auto-routes` default value in `sanitize_dhcp_state()`. It require extra code taking care of:
 * IP disabled, should apply all DHCP option as it might be copied from old profile.
 * DHCP disabled, should apply all DHCP option as it might be copied from old profile.

Currently code design is:
 * Clear all nmstate-known properties when duplicating old profile.
 * Only edit properties if required.

I merely follow the old design, if you think we should use another approach, another PR would be required.

EdDev commented:
> I merely follow the old design, if you think we should use another approach, another PR would be required.

I think we just have multiple solutions in placed for similar issues and we should converge to a single one if possible. That will make sure we have consistency and we can share this logic in nmstate core logic.

Anyway, I agree that this can be delayed to a refactoring PR.

cathay4t commented:
V7 patch uploaded.

Changes since V6:
 * Changes suggested by feedback.
 * Removed the patch which fix the Centos CI image as it already merged.
 * Split out the patch for fixing the incorrect use of mock in IPv4/IPv6 unit test.

cathay4t commented:
V8 patch uploaded. 

Changes since V7:
 * Renamed constants for classless route to `IPV4_CLASSLESS_ROUTE_DST_NET1`, `IPV4_CLASSLESS_ROUTE_NEXT_HOP1`, and `IPV6_CLASSLESS_ROUTE_DST_NET1`.
 * Changed the comment on `time.sleep()`. For IPv4, it's waiting DNS entry. For IPv6, it's waiting ipv6-RA and DHCPv6.
 * Changed the test case `test_ipv6_dhcp_only` to check IPv6 /128 address get from DHCPv6.

EdDev commented:
Thank you @cathay4t , looks great.
